### PR TITLE
Added proposal process.

### PR DESCRIPTION
**Pull request by bwplotka**
Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
Preview: https://deploy-preview-15--rhobs-handbook.netlify.app/proposals/done/202106-proposals-process.md/

**Review by ianbillett — COMMENTED** [review]
I'm a big fan of this approach - making big design docs public will only increase the visibility and accountability of our descisions.

In the spirit of this doc - it needs to be reviwed and approved by the leads so @squat & @simonpasquier

Also we should mention what make a confidential proposal? We talked yesterday about what makes a confidential doc: reference to specific customers, specific configurations etc. Is a confidential proposal any different? It could propose something that needs to be communicated with a partner perhaps? Worth being explicit.

**Comment by bwplotka** [comment]
Thanks for reivew @ianbillett 

Addressed comments, PTAL

**Comment by PhilipGough** [comment]
LGTM - I see this move as a positive thing. We tend to live in Github for the most part and I think this makes collaboration easier/better. Also benefits by making decisions open and adding visibility.

**Review by ianbillett — APPROVED** [review]
I'm comfortable with these changes - but let's get @squat and @simonpasquier signoff too as team leads :+1:

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:88]: I am a bit concerned with having only leads and architects as reviewers. That amounts to a pool of 3 people out of ~20 who can add feedback and approve changes. It also inhibits knowledge sharing and collaboration within the team since other members will be left out.

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:47]: How will the new process ensure that people drive their proposals to completion?

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:88]: Is there an SLO on how long we want proposals to:
* Stay in Review state
* Stay in Approved state before starting implementation.

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:70]: I would rename the subtitle to something similar to `Where to propose changes/Where to submit proposals`. I had to read the entire document 2-3 times to understand the difference between `What Process to Follow` and `Handbook Proposal Process`.

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:112]: Does this refer to upstream projects?

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:1]: Shouldn't this document be in the `Proposals/In Review` folder?

**Review by fpetkovski — COMMENTED** [review]
It would be nice to link the relevant documents, such as the google docs and markdown templates, in a References section.

Otherwise, looks quite solid so far.

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:1]: No, we do that only for Google Docs. For markdowns we know it's in progress since it's in PR (: WDYT?

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:47]: Very solid point. It's always based on authors motivation. Usually, the problem is not author motivation but lack of process, lack of visibility or decision making time. I think this proposal fixes those things.

It does not fix the motivation part. If author does not want to finish the proposal, maybe it's not that important and it will get just stale PR, closed after some time of inactivitiy

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:70]: Yessss

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:88]: I agree, was thinking about this too. Everyone should review. But question is, who is making final decision? What is the final vote? I would be totally for equal vote priority, but this would mean you have to have 11 out of 20 ppl to review which is unrealistic too. So at then end lead and architects has to know about it no?

This is BTW rare: What it will come next is that I will propose same to CMO, Observatorium, Prometheus Operator, Prometheus, Thanos etc, and then normal procedure applies as to any bigger change (consensus in maintainers team)

Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:88]: I don't think so. It's similar to PR (code change) process now? Depends on sprint priority. Do you think we need SLO?

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:112]: yesss

**Review by bwplotka — COMMENTED** [review]
Amazing  feedback, thanks @fpetkovski let's discuss

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:88]: Any proposed flow? Just get it reviewed by interested team members you choose? Consensus in the team?

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [reply content/Proposals/Done/202106-proposals-process.md:1]: Ah got it, makes complete sense :+1:

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [reply content/Proposals/Done/202106-proposals-process.md:90]: I think it makes complete sense to have the architect and lead from the team responsible for the topic as reviewers. 

I would, however, also encourage involving team members with the subject-specific domain knowledge in the review process. We have a list of contributors in [this doc](https://docs.google.com/spreadsheets/d/1plCB4Q-IB16_LhFiZiiZ3F9LPT-lHraphcm8D8Wusd8/edit#gid=1467586438) that we can use to select the appropriate team members.

In addition to that, anyone should feel free to comment on the proposal, regardless of whether they're selected as reviewers or not. In my view, a proposal should be treated as a tool for collaboration, innovation and dissipating knowledge within the team.

Comment (squat) [reply content/Proposals/Done/202106-proposals-process.md:90]: Review should be open to everyone, even outside of the team: this is open-source after all and anyone can comment, which is free input for us.

We _do_ want some accountability when it comes to approval to ensure that things do not hang around in limbo indefinitely and ensure that proposals are either accepted or rejected. In a regular team sync a few weeks ago I originally recommended that we follow a structured process like this inspired by Hashicorp's RFC processe [0]:

> Once you’ve written the first draft of an RFC, share it with your team. They’re likely to have the most context on your proposal and its potential impacts, so most of your feedback will probably come at this stage. Any team member can comment on and approve an RFC, but you need explicit approval only from the appropriate team leads in order to move forward. Once the RFC is approved and shared with stakeholders, you can start implementing the solution. For major projects, also share the RFC to the company-wide email list. While most members of the mailing list will just read the email rather than the full RFC, sending it to the list gives visibility into major decisions being made across the company. 

In our case, the architect would be ultimately responsible for signing off on something, but ANYONE can give feedback. We can even say: X people are chosen to review and give expert input on every document and present recommendations to the architect who ultimately signs off (or not). Alternatively, these X people could approve/reject the document by consensus.

[0]

File: content/Proposals/Done/202106-proposals-process.md
Comment (squat) [reply content/Proposals/Done/202106-proposals-process.md:47]: In my opinion, the guarantee of a process that says "your doc will be reviewed by a human in X time and you will be given feedback" is a good motivation. This, coupled with "your work is recognized and you can add it to your list of achievements" is much more than we currently have and should motivate the completion of designs more than we already do. The goals is not to ensure 100% completion of a design doc; it's impossible and not part of anyone's job description. But a structures process should create a positive difference IMO

File: content/Proposals/Done/202106-proposals-process.md
Comment (squat) [reply content/Proposals/Done/202106-proposals-process.md:90]: I think there should be an SLO for how long things stay in review state but not necessarily how long before they are implemented. Our problem to-date has been that docs hang around in review-limbo for eternity, which is super demotivating for writers and means they don't get credit for their good ideas. Whether or not a design is implemented is a different problem: the design may be a great idea but the feature may not be prioritized by the PM so it's not on our roadmap for the coming X quarters. Architects and leads should be responsible for circulating designs to the right ppl, e.g. managers and PMs, and to advocate for prioritizing them as needed, but it's not only up to the team to decide what we work on. Otherwise, cynically, this SLO would be a backdoor to override the needs of Red Hat as a business and get engineer hours dedicated to pet projects that have architect approval.

File: content/Projects/Observability/RHOBS/telemeter.md
Comment (squat) [inline content/Projects/Observability/RHOBS/telemeter.md:1]: This doesn't seem like it should be part of this PR

File: content/Projects/Observability/RHOBS/README.md
Comment (squat) [inline content/Projects/Observability/RHOBS/README.md:1]: This doesn't seem like it should be part of this PR

File: content/Projects/Observability/thanos.md
Comment (squat) [inline content/Projects/Observability/thanos.md:3]: This doesn't seem like it should be part of this PR

Comment (bwplotka) [reply content/Projects/Observability/thanos.md:3]: It is, just to show how to point to sub project proposal and show how project template could look on Thanos example.

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:95]: Yes, we have that part of the `TL;DR` I will add your words into that part for clarity (:

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:90]: Added mention to that to our process

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:47]: Added summary

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:90]: What specific SLO we should set?

Each iteration between change and review should not last longer than 3 days? 🤔

Comment (fpetkovski) [reply content/Proposals/Done/202106-proposals-process.md:90]: I think it makes sense to set something that is simple to start with. We can always adjust it if we need to. E.g. we can say each proposal needs to be approved or rejected within 1 month. We can also easily measure this SLO to see whether we are breaching it. 

I guess the more important question is, what is the SLA for breaching the SLO :)

Comment (PhilipGough) [reply content/Proposals/Done/202106-proposals-process.md:90]: @fpetkovski I guess if you breach the SLO then you lose the chance to voice your opinion and steer the proposal. It's a tricky one though because different proposals are going to have different scope and so might need lots/little time and or energy investment to review. Setting a low value for an SLO could be problematic in respect to timing for someone that might be interested in reviewing and have good opinions but be swamped etc?

I suppose at the very least if we do set an SLO, a comment from an interested party such as `I'm interested but don't have scope to review right now, I need more time!` would suffice, right?

Comment (fpetkovski) [reply content/Proposals/Done/202106-proposals-process.md:90]: That's a good point indeed, makes total sense :+1:

Comment (squat) [reply content/Proposals/Done/202106-proposals-process.md:90]: > I guess if you breach the SLO then you lose the chance to voice your opinion and steer the proposal

I'm thinking more about this from the other point of view, i.e. from that of the person proposing the document. If the SLO is breached, then it's them who suffer because their ideas won't be recognized. I think that in the last year this has been a bigger problem for the team than a reviewer not getting a say in the process.

Comment (PhilipGough) [reply content/Proposals/Done/202106-proposals-process.md:90]: @squat that's a fair point. I was mostly thinking that in a team of ~20, the proposer should be able to get some reviews within that time frame and drive the idea forward without being blocked by someone in particular.

So then is the suggestion that whoever proposes the idea might be able to identify and cc one or more people and push the onus on them to review within a specific timeframe?

Comment (squat) [reply content/Proposals/Done/202106-proposals-process.md:90]: I think that there needs to be some "official" reviewer who is accountable for approving or rejecting a proposal within x time. The problem IMO has not been that no one reads proposals but rather that they are read and they are never officially taken out of limbo. Even if you ping x person to review a doc, who is to say they are authoritative enough to give official sign-off on a proposal? For this reason it's tempting to default to saying architect/lead has to accept/reject a proposal within x weeks.

Comment (PhilipGough) [reply content/Proposals/Done/202106-proposals-process.md:90]: @squat yeah I get your point, makes total sense and at the end of the day someone or some relativity small group need to take the lead and "steer the ship" so I am all for it in that respect.

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [inline content/Proposals/Done/202106-proposals-process.md:98]:
```suggestion
As you see on the above algorithm, if the content relates to any upstream project, it should be proposed, reviewed and potentially implemented together with the community. This does not mean that you cannot involve other team members towards this effort. Share the proposal with team members, even if they are not part of maintainer's team on a given project, any feedback, and voice are useful and can help to move idea further.
```

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [reply content/Proposals/Done/202106-proposals-process.md:90]: > The problem IMO has not been that no one reads proposals but rather that they are read and they are never officially taken out of limbo.

I agree that this is a problem, I am just not sure if approving a proposal and not acting on it makes things better. With the current team processes, sprint and roadmap planning is done by a handful of people. This makes it hard for proposal authors to get visibility into if and when their proposal would be planned for implementation. Approving a proposal and not implementing it for a year or two still feels like a limbo state.

It was mentioned above that approving and implementing a proposal should be treated as different problems. I think this could be a good way to look at things. However, not having a feedback loop for the implementation phase leaves a big gap in the process.

File: content/Proposals/Done/202106-proposals-process.md
Comment (PhilipGough) [reply content/Proposals/Done/202106-proposals-process.md:98]: Alt: `This does not prevent you from involving other team members towards this effort`

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:90]: I totally see this problem, but I think it's better to NOT couple the proposal process with the implementation process.

> I am just not sure if approving a proposal and not acting on it makes things better.

I have to disagree. (: Let's say we don't have time to implement a new idea which someone has in mind. Do we think it will help to NOT do the proposal at least? I feel it's extremely helpful to at least cover that idea while it's hot, so anyone can take it up anytime later. Especially when we are working open-source it helps the community to get it up for us too! Not mentioning we would have a ready answer if someone asks about new idea that was discussed before. We can point how we would discuss this and that person can either update proposal if needed with new info or see what was decided before.

Also we do the same for bugs no? We acknowledge them but sometimes we can't fix everything. Still it's good to report them instead of ignoring and move the fix process as much as further as possible.

> However, not having a feedback loop for the implementation phase leaves a big gap in the process.

In some way there is a process for planning things for our teams: This is part of backlog grooming, sprint planning, syncs with team leads. Those are not properly documented, and probably there should be another piece of documentation for that WDYT? (:

File: content/Proposals/Done/202106-proposals-process.md
Comment (bwplotka) [reply content/Proposals/Done/202106-proposals-process.md:98]: hah (: Thanks!

File: content/Proposals/Done/202106-proposals-process.md
Comment (fpetkovski) [reply content/Proposals/Done/202106-proposals-process.md:90]: Ok, this seems like a good direction. To summarize:
* Keep approval and implementation phases independent
* Potentially iterate on adding visibility into planning and grooming processes of the teams. I would not rush this work to much before the team stabilizes. Constantly updating documentation can be a bit wasteful :)

Comment (paulfantom) [reply content/Proposals/Done/202106-proposals-process.md:90]: Proposal - follow KEP process.
- Proposal itself should state who is the approver/reviewer (creator can assign on first draft and can be changed in further iterations)
- Proposal should state if it is implementable right now or if it isn't
- Proposals without any activity are labeled as `stale` after 60d of inactivity
- Stale proposals are closed after an additional 30d of inactivity

More in KEP-0000 - https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md

**Pull request merged**
